storm-server/src/main/java/org/apache/storm/localizer/AsyncLocalizer.java

```diff
@@ -156,17 +146,141 @@ public AsyncLocalizer(Map<String, Object> conf, AtomicReference<Map<Long, LocalA
+        PortAndAssignment pna = new PortAndAssignment(port, assignment);
+        final String topologyId = assignment.get_topology_id();
+
+        LocallyCachedBlob topoJar = getTopoJar(topologyId);
+        topoJar.addReference(pna, cb);
+
+        LocallyCachedBlob topoCode = getTopoCode(topologyId);
+        topoCode.addReference(pna, cb);
+
+        LocallyCachedBlob topoConf = getTopoConf(topologyId);
+        topoConf.addReference(pna, cb);
+
+        CompletableFuture<Void> ret = topologyBasicDownloaded.get(topologyId);
+        if (ret == null) {
+            ret = downloadOrUpdate(topoJar, topoCode, topoConf);
+        }
+        return ret;
+    }
+
+    private static final int ATTEMPTS_INTERVAL_TIME = 100;
+
+    private CompletableFuture<Void> downloadOrUpdate(LocallyCachedBlob ... blobs) {
+        CompletableFuture<Void> [] all = new CompletableFuture[blobs.length];
+        for (int i = 0; i < blobs.length; i++) {
+            final LocallyCachedBlob blob = blobs[i];
+            all[i] = CompletableFuture.runAsync(() -> {
+                LOG.debug("STARTING download of {}", blob);
+                try (ClientBlobStore blobStore = ServerUtils.getClientBlobStoreForSupervisor(conf)) {
+                    boolean done = false;
+                    long failures = 0;
+                    while (!done) {
+                        try {
+                            synchronized (blob) {
+                                long localVersion = blob.getLocalVersion();
+                                long remoteVersion = blob.getRemoteVersion(blobStore);
+                                if (localVersion != remoteVersion) {
+                                    try {
+                                        long newVersion = blob.downloadToTempLocation(blobStore);
+                                        blob.informAllOfChangeAndWaitForConsensus();
+                                        blob.commitNewVersion(newVersion);
+                                        blob.informAllChangeComplete();
+                                    } finally {
+                                        blob.cleanupOrphanedData();
+                                    }
+                                }
+                            }
+                            done = true;
+                        } catch (Exception e) {
+                            failures++;
+                            if (failures > blobDownloadRetries) {
+                                throw new RuntimeException("Could not download...", e);
+                            }
+                            LOG.warn("Failed to download blob {} will try again in {} ms", blob, ATTEMPTS_INTERVAL_TIME, e);
+                            Utils.sleep(ATTEMPTS_INTERVAL_TIME);
+                        }
+                    }
+                }
+                LOG.debug("FINISHED download of {}", blob);
```

Comment:
Should this be info?

Reply:
I used it for debugging.  If you see a good reason for it to be info I can make it info, but I didn't see a lot of value in it.